Modules/sha512module.c

```diff
@@ -460,12 +463,13 @@ SHA512Type_copy_impl(SHAobject *self)
 /*[clinic end generated code: output=adea896ed3164821 input=9f5f31e6c457776a]*/
 {
     SHAobject *newobj;
+    SHA512State *st = PyType_GetModuleState(Py_TYPE(self));
```

Comment:
This code is illegal if you are not sure that self is an instance of a SHA512 object. You should get the defining class using Argument Clinic instead. Something like "cls: defining_class", see Modules/posixmodule.c for example.